src/rpc/daemon_rpc_client.cpp

```diff
@@ -0,0 +1,623 @@
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "rpc/daemon_rpc_client.h"
+#include "serialization/json_object.h"
+
+namespace cryptonote
+{
+
+namespace rpc
+{
+
+DaemonRPCClient::DaemonRPCClient()
+{
+}
+DaemonRPCClient::~DaemonRPCClient()
+{
+}
+
+void DaemonRPCClient::connect(const std::string& addr, const std::string& port)
+{
+  zmq_client.connect(addr, port);
+}
+
+bool DaemonRPCClient::getHeight(uint64_t& height)
+{
+  rapidjson::Value response_json;
+
+  try
+  {
+    GetHeight::Request request;
+
+    response_json = doRequest<GetHeight>(request);
+  }
+  catch (...)
+  {
+    return false;
+  }
+
+  try
+  {
+    GetHeight::Response response = parseResponse<GetHeight>(response_json);
+
+    height = response.height;
+
+    return true;
+  }
+  catch (...)
+  {
+  }
+
+  return false;
+}
+
+bool DaemonRPCClient::getBlocksFast(
+    const std::list<crypto::hash>& block_ids,
+    const uint64_t start_height_in,
+    std::vector<cryptonote::rpc::block_with_transactions>& blocks,
+    uint64_t& start_height_out,
+    uint64_t& current_height,
+    std::vector<cryptonote::rpc::block_output_indices>& output_indices,
+    std::string& error_details)
+{
+  rapidjson::Value response_json;
+
+  try
+  {
+    GetBlocksFast::Request request;
+
+    request.block_ids = block_ids;
+    request.start_height = start_height_in;
+
+    response_json = doRequest<GetBlocksFast>(request);
+  }
+  catch (...)
+  {
+    return false;
+  }
+
+  try
+  {
+    GetBlocksFast::Response response = parseResponse<GetBlocksFast>(response_json);
+
+    blocks = response.blocks;
+    start_height_out = response.start_height;
+    current_height = response.current_height;
+    output_indices = response.output_indices;
+
+    return true;
+  }
+  catch (...)
+  {
+    try
+    {
+      cryptonote::rpc::error err = parseError(response_json);
+
+      error_details = err.error_str;
+    }
+    catch (...)
+    {
+      error_details = "Daemon returned improper JSON-RPC response.";
```

Comment:
Both, actually.  The transport layer is handled by 0MQ, but the messages passed are JSON-RPC compliant.